Add JAX implementation fol `MatrixIsPositiveDefinite` `Op`

juanitorduz opened the pull request:
Closes: https://github.com/pymc-devs/pymc/issues/6849

Following https://pytensor.readthedocs.io/en/latest/extending/creating_a_numba_jax_op.html

<!-- readthedocs-preview pymc start -->
----
:books: Documentation preview :books:: https://pymc--6853.org.readthedocs.build/en/6853/

<!-- readthedocs-preview pymc end -->

Commit:
Remove SpecifyShape from Assert JAX

juanitorduz commented on tests/sampling/test_jax.py:
```diff
@@ -49,6 +52,14 @@ def test_old_import_route():
+    matrix_pos_def = PosDefMatrix()
+    x_is_pos_def = matrix_pos_def(x)
+    out_fg = FunctionGraph(outputs=[x_is_pos_def])
+    compare_jax_and_py(fgraph=out_fg, test_inputs=pt.eye(3))
```
@ricardoV94 I can not use `from tests.link.jax.test_basic import compare_jax_and_py` from `pymc` (because it belongs to the tests) ...shall we add a copy?

ricardoV94 replied:
I don't think there's anything special about that. I would just compile a function to JAX and confirm it returns the expected results

juanitorduz replied:
I want to get the test working with a simple case, and then I will add parametrize it with edge cases.

ricardoV94 replied:
We should compile only once and test with multiple values (pos and not pos). Parametrize is not ideal because we don't want to recompile/jit just to test a new value

juanitorduz commented on pymc/sampling/jax.py:
```diff
@@ -74,6 +74,15 @@ def assert_fn(value, *inps):
+def jax_funcify_PosDefMatrix(op, **kwargs):
+    def posdefmatrix_fn(value, *inps):
+        no_pos_def = jnp.any(jnp.isnan(jnp.linalg.cholesky(value)))
+        return (no_pos_def == jnp.array(False, dtype="bool")).astype(jnp.int_)
```
Do we have a better way to negate a statement and convert it to `int`? This smells bad ...

ricardoV94 replied:
The PyTensor Op doesn't return a bool? It seems like it should

juanitorduz replied:
It returns an integer:

```python
import pytensor.tensor as pt
from pymc.distributions.multivariate import PosDefMatrix

matrix_pos_def = PosDefMatrix()

matrix_pos_def(pt.eye(2)).eval()
>> array(1, dtype=int8)
```

ricardoV94 replied:
Let's check if we can change it to bool. Should need tweaking of `make_node` and `perform` of the original Op

juanitorduz replied:
Tried to do it in https://github.com/pymc-devs/pymc/pull/6853/commits/013ead3c8b00b7331429f7702a90092ea733e1a7

juanitorduz replied:
(and tests https://github.com/pymc-devs/pymc/pull/6853/commits/75ca38bdc7f1bf4acee903cfec8b79de3bfa5f10)

ricardoV94 replied:
Then the jax one can/shoud return the boolean variable `not any(...)`

juanitorduz replied:
In https://github.com/pymc-devs/pymc/pull/6853/commits/b9a2ef8007bfcce9c166858addf909b3c3bce878 I made the output boolean (as in the tests)

ricardoV94 commented on pymc/sampling/jax.py:
```diff
@@ -74,6 +74,15 @@ def assert_fn(value, *inps):
+def jax_funcify_PosDefMatrix(op, **kwargs):
+    def posdefmatrix_fn(value, *inps):
+        no_pos_def = jnp.any(jnp.isnan(jnp.linalg.cholesky(value)))
+        return no_pos_def == jnp.array(False, dtype="bool")
```
Isn't this enough?

```suggestion
        return not no_pos_def
```

ricardoV94 replied:
I would make it a one liner because the double negation with the name is confusing

ricardoV94 commented on tests/sampling/test_jax.py:
```diff
@@ -49,6 +51,26 @@ def test_old_import_route():
+        (jnp.array([[1, 0.99, 1], [0.99, 1, 0.999], [1, 0.999, 1]], dtype="float32"), False),
+    ],
+)
+def test_jax_PosDefMatrix(input, expected):
```
Let's not parametrize, it's quite wasteful to compile a function for each different value we want to test.

juanitorduz replied:
So should l loop over the cases inside the test to test all the cases? (after the compilation)

ricardoV94 replied:
Yup

ricardoV94 committed:
Fix TestMatrixNormal.check_draws

Test behavior was accidentally changed in 9dad9c2baeef9d19c7cd7ff740e30843e66df93e

Commit:
Make PosDefMatrix Op output type boolean

Commit:
Add JAX implementation of PosDefMatrix Op

ricardoV94 committed:
Seed flaky TestHSGP.test_prior

juanitorduz commented on tests/gp/test_hsgp_approx.py:
```diff
@@ -178,7 +178,7 @@ def test_prior(self, model, cov_func, X1, parameterization):
             f2 = gp.prior("f2", X=X1)
 
-            idata = pm.sample_prior_predictive(samples=1000)
+            idata = pm.sample_prior_predictive(samples=1000, random_seed=rng)
```
🙏

ricardoV94 replied:
One day we will get this 5 line PR merged. Just you wait :D
